### DISCUSSION TITLE
Environment Variable undefined at build time

### DISCUSSION THREAD

**Discussion by doug-skinner** [vercel/next.js#38164] [Help] [answered]
Hello! I am running into an issue with using `getServerSideProps` instead of an API route called from the client side to load data. 

Example files:

```javascript
// src/config.js
export default {
    oktaDomain: process.env.OKTA_DOMAIN,
    oktaToken: process.env.OKTA_TOKEN,
};
```

```javascript
// src/services/clients/oktaClient.js
import okta from '@okta/okta-sdk-nodejs';
import config from '~/config';

const _getClient = () => {
    const client = new okta.Client({
        orgUrl: config.oktaDomain,
        token: config.oktaToken,
    });
    return client;
};

const client = _getClient();

export default client;
```

```javascript
// src/services/okta.js
import client from './clients/oktaClient';

export const getUser = async (userId) => {
    const user = await client.getUser(userId);
    return user;
};
```

```javascript
// src/pages/users/[userId].js
import { getUser } from '~/services/okta';

const UserDetails = ({ user }) => (<div>Hi {user.name}!</div>);
export default UserDetails;

export const getServerSideProps = async ({ params }) => {
    const { userId } = params;
    const user = await getUser(userId);
    return { props: { user } };
};
```

When I was calling the `await getUser(userId)` from an API route, which runs on the server side, `next build` worked perfectly fine, built the app, the value of the environment variables was looked up at run time and everything is golden. 

Now though, with the call moved into `getServerSideProps`, I'm getting a build error:

```
#13 15.09 > Build error occurred
#13 15.09 Error: Found 2 errors:
#13 15.09 Okta Org URL not provided
#13 15.09 Okta API token not provided
#13 15.09 See https://github.com/okta/okta-sdk-nodejs for usage.
#13 15.09     at new Client (/app/node_modules/@okta/okta-sdk-nodejs/src/client.js:65:13)
#13 15.09     at _getClient (/app/.next/server/chunks/298.js:30:21)
#13 15.09     at Object.6298 (/app/.next/server/chunks/298.js:36:16)
#13 15.09     at __webpack_require__ (/app/.next/server/webpack-runtime.js:25:42)
#13 15.09     at Object.4966 (/app/.next/server/pages/groups/[groupId].js:28:72)
#13 15.09     at __webpack_require__ (/app/.next/server/webpack-runtime.js:25:42)
#13 15.09     at __webpack_exec__ (/app/.next/server/pages/groups/[groupId].js:285:39)
#13 15.09     at /app/.next/server/pages/groups/[groupId].js:286:86
#13 15.09     at Function.__webpack_require__.X (/app/.next/server/webpack-runtime.js:190:21)
#13 15.09     at /app/.next/server/pages/groups/[groupId].js:286:47 {
#13 15.09   type: 'Error'
#13 15.09 }
```

It looks like Next is trying to look up the variables which aren't populated yet and then making them `undefined`, which breaks creating the client in `src/services/clients/okta.js` above. 

What I don't understand though is why it is trying to look up the environment variables at build time rather than at run time for code that is running on the server. I understand that [according to the docs](https://nextjs.org/docs/basic-features/environment-variables#loading-environment-variables) it tries to do this, but I don't understand why when it doesn't do this for API routes.

Can someone help me out or suggest an alternative? I don't want to go back to client side loading, as I should be able to use server side props for this.

Thanks in advance!

**Comment by icyJoseph** [comment]
I'll try to write up on what I think is happening.

In the mean time, please use runtime config https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration, `serverRuntimeConfig`, specifically.

### Edit

I just did some testing, and this works very well. I build with `yarn build`, and then run my server with `FOO=bar npx next start` and in GSSP and API routes, the value of `process.env.FOO` is 'bar'. If I restart, without rebuilding, `FOO=buzz npx next start`, then `process.env.FOO` is 'buzz'.

Is there anything else that could be missing in your snippets?

**Reply by doug-skinner** [reply]
Thank you, that change did fix this, although I am very curious to see what your longer writeup is about this!

**Reply by icyJoseph** [reply]
I did a small edit on my answer.

However, I've seen this in one of our deploys, where variables were undefined, suddenly. We use docker to deploy the app that has that issue.

**Reply by doug-skinner** [reply]
Yea I actually spoke too soon, this is still not working with a build within Docker...

**Reply by doug-skinner** [reply]
And the Dockerfile just copies in the necessary files and runs `npm run build`. Which is what gives the error message seen above. I can build successfully outside of Docker though, which is weird. And the same exact config builds fine within Docker when the code is called through an API route rather than through server side props.

**Reply by doug-skinner** [reply]
More info, when logging the `serverRuntimeConfig` during build time I am getting an empty object, which also seems wrong.

**Reply by icyJoseph** [reply]
Oh no, docker issues... Are all env vars present in the docker file? 

Could you share a bit of what happens in the docker file?

Maybe this SO answers help? https://stackoverflow.com/a/49045243

> The environment variables not declared inside the Dockerfile are not visible to the container when building the image. They are only passed to the container at runtime.

Somehow this rings a bell, but still, why would api routes and GSSP behave differently, GSSP is almost just a specialized api route that pipes its result into a component, or sends it over the network as JSON.

**Reply by doug-skinner** [reply]
I’ll share the full docker file tomorrow with access to the computer the code is on, but I can say none of the environment variables are present in it. We run the container on EKS and provide the environment variables there. Which is the main reason why I want to use the environment variables because they change per environment. And the same docker file without the variables does work when using the api routes. I’m definitely a little confused at this point.

**Reply by doug-skinner** [reply]
Okay, here's the Dockerfile, based on one of the next.js example repos:

```Dockerfile
# Install dependencies only when needed
FROM node:16-alpine AS deps

WORKDIR /app
COPY package.json package-lock.json ./
RUN --mount=type=secret,id=npmrc,dst=./.npmrc npm ci --production

# Rebuild the source code only when needed
FROM node:16-alpine AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .
RUN npm run build

# Procution image, copy all the files in and run the application
FROM node:16-alpine AS runner
WORKDIR /app

ENV NODE_ENV=production

RUN addgroup -g 1001 -S nodejs
RUN adduser -S nextjs -u 1001

COPY --from=builder /app/next.config.js ./
COPY --from=builder /app/public ./public
COPY --from=builder /app/package.json ./package.json
COPY --from=builder /app/node_modules ./node_modules/

# Automatically leverage output traces to reduce image size 
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000

CMD ["npm", "run", "start"]
```

It fails on the `npm run build` command. That same command works outside of the docker container, even if the environment variables are empty. I'm gonna keep debugging this, but wanted to throw this here.

**Reply by GabenGar** [reply]
Do you have `.env` files in `.dockerignore`? How do you even pass env variables into the images in the first place?

**Reply by doug-skinner** [reply]
There are no environment files being passed to Docker, and the build works on my machine even without any `.env` files. The environment variables are passed at runtime to the docker container running on EKS. The whole thing builds when not using `getServerSideProps`, but when that is used it tries to evaluate the env variables even though that is supposed to be server side code, not client side.

**Reply by GabenGar** [reply]
Well yeah env vars are string-replaced at build time, as in `process.env.*` are treated as string literals for replacement as far as code bundler concerned. You can't use them at runtime without passing at build time.

**Reply by doug-skinner** [reply]
For what code though? I've been doing things exactly as described for 3 years with `process.env` through various API endpoints. It's only when now trying to use `getServerSideProps` that this is coming up. According to the docs, any code in `getServerSideProps` should be treated as server side and not built for the frontend, but clearly it's trying to build it for the frontend as you're describing contrary to the documentation.

**Reply by GabenGar** [reply]
> I've been doing things exactly as described for 3 years with process.env through various API endpoints.

Were you doing it with nextjs or not?

> According to the docs, any code in getServerSideProps should be treated as server side and not built for the frontend, but clearly it's trying to build it for the frontend as you're describing contrary to the documentation.

That's not about serverside or not, its about bundling and nextjs does bundle server code too. Which means all `process.env.*` calls are replaced with the literal values of `process.env.*` at build time. It says as much in the [env var docs](https://nextjs.org/docs/basic-features/environment-variables#loading-environment-variables).

**Reply by doug-skinner** [reply]
Yes, using Nextjs for the last 3 years. As I said above, I have api routes within Nextjs in this same application that work just fine. This is only an issue when using `getServerSideProps` inside a page. 

All of the code in my api routes that use `process.env.*` are **not** being replaced at build time. They are using the environment variables populated at runtime. That is the entire point of environment variables, to change per environment.

**Comment by doug-skinner** [comment, marked as answer]
This has been worked around with the following change to `/ src/services/clients/oktaClient.js`:

```javascript
import okta from '@okta/okta-sdk-nodejs';
import config from '~/config';

let client;

export const getClient = () => {
    if (client) {
        return client;
    }

    client = new okta.Client({
        orgUrl: config.oktaDomain,
        token: config.oktaToken,
    });

    return client;
};
```

And then calling `getClient` within the files used by serverside props. Because the code no longer runs at build time, everything is peachy. Thanks everyone for your help

**Reply by icyJoseph** [reply]
Aha, a good ol' closure. I'll try to find time and give this a go on our application. Runtime configs did the trick for us, but I still wonder what the root issue is. It itches that there's no clear explanation, other than, the variables were missing at build time.

Are you perhaps importing part of this code to client code? or something like that? Ah, perhaps is best to let this issue rest.

**Reply by doug-skinner** [reply]
I completely agree that this answer is not satisfying. My best guess at what was going on is even though the documentation says that all imports within pages that are only used within `getServerSideProps` shouldn't be built for the client page and instead should be extracted into a "fake api route" behind the scenes, this wasn't happening for some reason. Which is weird and shouldn't happen.

Either way, thanks for your help the other day and for re-assuring me that this is in fact a really weird thing that's happening.
